### ISSUE TITLE
Previous curve type not kept for floats when resizing animation

### ISSUE THREAD

**Issue by via5** [acidbubbles/vam-timeline#42]
1) In a new animation, add a controller target (such as `headControl`) and a float parameter (such as `morph: Mouth Open`).
2) Uncheck "Loop", keep length at 2s.
3) Move the scrubber to 1s.
4) Add a keyframe for both the controller and the float parameter.
5) Change the length to 3s.

Notice that the controller does not have a keyframe at 2s, but the float parameter does, carried over from the last, pseudo keyframe in the previous animation length. I suspect there's something in 50cf271e89041259f367eb7117b9e45e9628b981 that doesn't handle float parameters the same way as controllers. There should not be any keyframes at 2s.

**Comment by acidbubbles** [comment]
Ah yeah I see, I made it so that if your last frame uses CopyPrevious it didn't keep the useless previous-to-last keyframe but that's only for controllers. I'll check for floats and triggers too!

**Comment by via5** [comment]
Just to confirm, the test case above still fails as before, even with the commit you did.

But now I also found this case:

1) In a new animation, add a controller target (such as `headControl`) and a float parameter (such as `morph: Mouth Open`).
2) Leave "Loop" _checked_, keep length 2s.
3) Move the scrubber to 1s.
4) Add a keyframe for both the controller and the float parameter.
5) Uncheck "Loop".
6) Change length to 3s.

Both `headControl` and the morph will have a keyframe at 2s. They shouldn't.

The main problem is that once you've encountered either case, this specific animation loses the CopyPrevious state permanently. You either have to put up with a last keyframe that's gets farther and farther away from where you're working until you're done animating, or you have to start over.

**Comment by acidbubbles** [comment]
It does? Argh, I'm sure I verified that it worked :| Thanks, I'll check again, maybe I just fixed another variation of this issue :|

**Comment by acidbubbles** [comment]
So the behavior is normal. Resizing an animation with more than two keyframes will keep all existing keyframes (this avoid some unexpected stretching). However I think I found a fairly elegant solution, now if you press Delete on the last keyframe on a non-looping clip, it'll change the curve type to "copy previous" (which is the same as deleting in practice). Let me know if you think this is good enough to close this issue, I pushed it on `master` already!

**Comment by via5** [comment]
I'm not sure I understand. In my example above, I didn't add a keyframe at 2s. Why would resizing add one?

**Comment by acidbubbles** [comment]
If your animation is 2s, it means you will _always_ have a keyframe at 0s (zero) and 2s (length). If you move from a loop to a non-loop animation, it will not "destroy" your last frame, so you have three keyframes. The only case where it will destroy your last frame (and use copy previous) if is you have _two_ keyframes, since two keyframes in a looping animations means there's nothing to destroy. Other cases are full of edge cases where deleting the last frame before extending can create confusion. In this case for example, imaging the animation is a head saying no. 0s looks left, 1s look right, 2s is the loop point. When you swith to non-looping, you still have the "no" animation, though the last frame is now free to be modified as you wish. If you extend the animation by 2s, what it'll do now is keep the head saying no as is, and add 2s to it. If I were to delete the 1s keyframe or move it, you'd get one head direction right, and the other would just slow down. That's just one of the situations, you can have some targets with 4, 5 keyframes and others with 3, they would become out of sync if I were to just "move the last one".

So long story short:

1. Moving from looping to non-looping will use copy previous if there are only two keyframes.
2. If there are more, it will preserve the animation.
3. Extending an animation will keep the last keyframe.
4. To let you bring back the copy previous behavior on partially edited animations, I added the "delete" feature on the last keyframe.

Does this make more sense?

**Comment by via5** [comment]
Okay, well, the delete last takes care of my second case, it only happens when I forget to switch to non looping when I start anyway. Thank you for the explanation, I appreciate it.

My original test case still fails for float params though ;)

**Comment by acidbubbles** [comment]
Slippery one you! And I don't have excuses now, I _do_ have unit tests on this! I'll check this out again soon, thanks :P

**Comment by acidbubbles** [comment]
Fixed, _for real now_ (hopefully :D )
